File: nlsCppSdk/framework/nlsClient.cpp
/*
 * Copyright 2015 Alibaba Group Holding Limited
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
#include "Config.h"
#include "nlsClient.h"
#include "log.h"
#include "connectNode.h"
#include "eventNetWork.h"

#include "sr/speechRecognizerRequest.h"
//#include "sr/speechRecognizerSyncRequest.h"
#include "st/speechTranscriberRequest.h"
//#include "st/speechTranscriberSyncRequest.h"
#include "sy/speechSynthesizerRequest.h"
#include "da/dialogAssistantRequest.h"

#if !defined(__APPLE__)
#include "commonSsl.h"
#endif

namespace AlibabaNls {

using namespace utility;
//using namespace transport;

NlsClient* NlsClient::_instance = NULL;//new NlsClient();
bool NlsClient::_isInitializeSSL = false;
bool NlsClient::_isInitializeThread = false;
#if defined(_WIN32)
	HANDLE NlsClient::_mtx = CreateMutex(NULL, FALSE, NULL);
#else
	pthread_mutex_t NlsClient::_mtx = PTHREAD_MUTEX_INITIALIZER;
#endif


/*
 1: init openssl
 2: init log
 3: init vipServer
 4: startup libevent work threads
 */
NlsClient* NlsClient::getInstance(bool sslInitial) {

#if defined(_WIN32)
	WaitForSingleObject(_mtx, INFINITE);
#else
	pthread_mutex_lock(&_mtx);
#endif
	if (NULL == _instance) {
		//init openssl
    	if(sslInitial) {
        	if(!_isInitializeSSL) {
#if !defined(__APPLE__)
				SslConnect::init();
#endif
            	_isInitializeSSL = sslInitial;
        	}
    	}

		//init libevent work threads
//		NlsEventClientNetWork::initEventNetWork(count);

		//init nlsClient
		_instance = new NlsClient();

    }
#if defined(_WIN32)
	ReleaseMutex(_mtx);
#else
	pthread_mutex_unlock(&_mtx);
#endif

    return _instance;
}

/*
 1: destroy openssl
 2: destroy log
 3: destroy vipServer
 4: stop libevent work threads
 */
void NlsClient::releaseInstance() {
#if defined(_WIN32)
		WaitForSingleObject(_mtx, INFINITE);
#else
		pthread_mutex_lock(&_mtx);
#endif
    if (_instance) {
		LOG_DEBUG("release NlsClient.");

		if(_isInitializeThread) {
			NlsEventClientNetWork::destroyEventNetWork();
		}

		if (_isInitializeSSL) {
#if !defined(__APPLE__)
			LOG_DEBUG("delete NlsClient release ssl.");

			SslConnect::destroy();
#endif
			_isInitializeSSL = false;
		}

		NlsLog::destroyLogInstance();

        delete _instance;
        _instance = NULL;
    }

#if defined(_WIN32)
		ReleaseMutex(_mtx);
#else
		pthread_mutex_unlock(&_mtx);
#endif
}

NlsClient::NlsClient() {
//	_eventWork = NULL;
}

NlsClient::~NlsClient() {
//	delete _eventWork;
//	_eventWork = NULL;
}

const char* NlsClient::getVersion()	{
	return NLS_SDK_VERSION_STR;
}

void NlsClient::startWorkThread(int threadsNumber) {
#if defined(_WIN32)
	WaitForSingleObject(_mtx, INFINITE);
#else
	pthread_mutex_lock(&_mtx);
#endif

	if(!_isInitializeThread) {
		NlsEventClientNetWork::initEventNetWork(threadsNumber);
		_isInitializeThread = true;
	}

#if defined(_WIN32)
	ReleaseMutex(_mtx);
#else
	pthread_mutex_unlock(&_mtx);
#endif

}

int NlsClient::setLogConfig(const char* logOutputFile, const LogLevel logLevel, unsigned int logFileSize) {

	if ((logLevel < 1) || (logLevel > 4)) {
		return -1;
	}

	if (logFileSize < 0) {
		return -1;
	}

	NlsLog::_logInstance->logConfig(logOutputFile, logLevel, logFileSize);

	return 0;
}

void NlsClient::releaseRequest(INlsRequest* request) {
	if (request->getConnectNode()->getConnectNodeStatus() == NodeInitial) {
		LOG_INFO("released the SpeechRecognizerRequest");
		delete request;
		request = NULL;
		return ;
	}

	if (request->getConnectNode()->updateDestroyStatus()) {
		if (request->getConnectNode()->getConnectNodeStatus() == NodeInvalid) {
			LOG_INFO("released the SpeechRecognizerRequest");
			delete request;
			request = NULL;
			return;
		}
	}
}

SpeechRecognizerRequest* NlsClient::createRecognizerRequest() {
	return new SpeechRecognizerRequest();
}

void NlsClient::releaseRecognizerRequest(SpeechRecognizerRequest* request) {
	if (request) {
		if (request->getConnectNode()->getExitStatus() == ExitInvalid) {
			request->stop();
		}

		releaseRequest(request);
	}
}

SpeechTranscriberRequest* NlsClient::createTranscriberRequest() {
	return new SpeechTranscriberRequest();
}

void NlsClient::releaseTranscriberRequest(SpeechTranscriberRequest* request) {
	if (request) {
		if (request->getConnectNode()->getExitStatus() == ExitInvalid) {
			request->stop();
		}

		releaseRequest(request);
	}
}

SpeechSynthesizerRequest* NlsClient::createSynthesizerRequest(TtsVersion version){
	return new SpeechSynthesizerRequest((int)version);
}

void NlsClient::releaseSynthesizerRequest(SpeechSynthesizerRequest* request) {
	if (request) {
		if (request->getConnectNode()->getExitStatus() == ExitInvalid) {
			request->stop();
		}

		releaseRequest(request);
	}
}

DialogAssistantRequest* NlsClient::createDialogAssistantRequest(DaVersion version) {
    return new DialogAssistantRequest((int) version);
}

void NlsClient::releaseDialogAssistantRequest(DialogAssistantRequest* request) {
	if (request) {
		if (request->getConnectNode()->getExitStatus() == ExitInvalid) {
			request->stop();
		}

		releaseRequest(request);
	}
}


}

File: nlsCsharpSdk/nlsCsharpSdk/NlsClient.cs
/*
 * Copyright 2021 Alibaba Group Holding Limited
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

using nlsCsharpSdk.CPlusPlus;
using System.Runtime.InteropServices;

namespace nlsCsharpSdk
{
    /// <summary>
    /// NLS SDK 操作入口.
    /// </summary>

    public class NlsClient : INlsClient
    {
        #region Get the version of NLS SDK
        /// <summary>
        /// NLS SDK 当前版本信息.
        /// </summary>
        /// <returns></returns>
        public string GetVersion() 
        {
            IntPtr get = NativeMethods.NlsGetVersion();
            string version = "unknown version";
            if (get != IntPtr.Zero)
            {
                version = Marshal.PtrToStringAnsi(get);
            }
            return version;
        }
        #endregion

        #region Set the configure of log system
        /// <summary>
        /// 设置日志文件与存储路径.
        /// </summary>
        /// <param name="logOutputFile">
        /// 日志文件, 绝对路径或者相对路径均可. 若填写"log-transcriber", 则会在程序当前目录生成log-transcriber.log;
        /// 若填写"/home/XXX/NlsCppSdk/log-transcriber", 则会在/home/XXX/NlsCppSdk/下生成log-transcriber.log.
        /// </param>
        /// <param name="logLevel">
        /// 日志级别, 默认1 (LogError:1, LogWarning:2, LogInfo:3, LogDebug:4).
        /// </param>
        /// <param name="logFileSize">
        /// 日志文件的大小, 以MB为单位, 默认为10MB;如果日志文件内容的大小超过这个值,
        /// SDK会自动备份当前的日志文件, 超过后会循环覆盖已有文件.
        /// </param>
        /// <param name="logFileNum">
        /// 日志文件循环存储最大数, 默认10个文件.
        /// </param>
        /// <returns></returns>
        public int SetLogConfig(string logOutputFile, LogLevel logLevel, int logFileSize, int logFileNum)
        {
            return NativeMethods.NlsSetLogConfig(logOutputFile, (int)logLevel, logFileSize, logFileNum);
        }
        #endregion

        #region Set the socket address type
        /// <summary>
        /// 设置套接口地址结构的类型, 若调用需要在StartWorkThread之前.
        /// </summary>
        /// <param name="aiFamily">
        /// 套接口地址结构类型 AF_INET/AF_INET6/AF_UNSPEC
        /// </param>
        /// <returns></returns>
        public void SetAddrInFamily(string aiFamily)
        {
            NativeMethods.NlsSetAddrInFamily(aiFamily);
        }
        #endregion

        #region Set the ipv4 address
        /// <summary>
        /// 跳过dns域名解析直接设置服务器ip地址, 若调用则需要在StartWorkThread之前.
        /// </summary>
        /// <param name="ip">
        /// ipv4的地址 比如106.15.83.44
        /// </param>
        /// <returns></returns>
        public void SetDirectHost(string ip)
        {
            NativeMethods.NlsSetDirectHost(ip);
        }
        #endregion

        #region calculate the text
        /// <summary>
        /// 待合成音频文本内容字符数.
        /// </summary>
        /// <param name="text">
        /// 需要传入UTF-8编码的文本内容
        /// </param>
        /// <returns>
        /// 返回字符数
        /// </returns>
        public int CalculateUtf8Chars(string text)
        {
            return NativeMethods.NlsCalculateUtf8Chars(text);
        }
        #endregion

        #region Start WorkThread (Init NLS SDK)
        /// <summary>
        /// 启动工作线程数量, 同时也是NLS SDK的初始化步骤.
        /// </summary>
        /// <param name="threadsNumber">
        /// 启动工作线程数量, 默认设置值为 1.
        /// threadsNumber 小于 0 则启动工作线程数量等于CPU核数.
        /// </param>
        /// <returns></returns>
        public void StartWorkThread(int threadsNumber = 1)
        {
            NativeMethods.NlsStartWorkThread(threadsNumber);
            return;
        }
        #endregion

        #region Release NLS SDK
        /// <summary>
        /// 销毁NlsClient对象实例, 即释放NLS SDK.
        /// </summary>
        /// <returns></returns>
        public void ReleaseInstance()
        {
            NativeMethods.NlsReleaseInstance();
            return;
        }
        #endregion

        #region Creata a request of speech transcriber
        /// <summary>
        /// 创建实时音频流识别对象.
        /// </summary>
        /// <returns>成功返回SpeechTranscriberRequest对象, 否则返回NULL.</returns>
        public SpeechTranscriberRequest CreateTranscriberRequest()
        {
            IntPtr request = NativeMethods.NlsCreateTranscriberRequest();
            SpeechTranscriberRequest STrequest = new SpeechTranscriberRequest();
            STrequest.native_request = request;
            return STrequest;
        }
        #endregion

        #region Release the request of speech transcriber
        /// <summary>
        /// 销毁实时音频流识别对象.
        /// </summary>
        /// <param name="request">
        /// CreateTranscriberRequest所建立的request对象.
        /// </param>
        /// <returns></returns>
        public void ReleaseTranscriberRequest(SpeechTranscriberRequest request)
        {
            NativeMethods.NlsReleaseTranscriberRequest(request.native_request);
            request.native_request = IntPtr.Zero;
            return;
        }
        #endregion

        #region Creata a request of speech recognizer
        /// <summary>
        /// 创建一句话识别对象.
        /// </summary>
        /// <returns>成功返回speechRecognizerRequest对象, 否则返回NULL.</returns>
        public SpeechRecognizerRequest CreateRecognizerRequest()
        {
            IntPtr request = NativeMethods.NlsCreateRecognizerRequest();
            SpeechRecognizerRequest SRrequest = new SpeechRecognizerRequest();
            SRrequest.native_request = request;
            return SRrequest;
        }
        #endregion

        #region Release the request of speech recognizer
        /// <summary>
        /// 销毁一句话识别对象.
        /// </summary>
        /// <param name="request">
        /// CreateRecognizerRequest所建立的request对象.
        /// </param>
        /// <returns></returns>
        public void ReleaseRecognizerRequest(SpeechRecognizerRequest request)
        {
            NativeMethods.NlsReleaseRecognizerRequest(request.native_request);
            request.native_request = IntPtr.Zero;
            return;
        }
        #endregion

        #region Creata a request of speech synthesizer
        /// <summary>
        /// 创建语音合成对象.
        /// </summary>
        /// <param name="version">
        /// tts类型, 短文本或长文本.
        /// </param>
        /// <returns></returns>
        public SpeechSynthesizerRequest CreateSynthesizerRequest(TtsVersion version)
        {
            IntPtr request = NativeMethods.NlsCreateSynthesizerRequest((int)version);
            SpeechSynthesizerRequest SYrequest = new SpeechSynthesizerRequest();
            SYrequest.native_request = request;
            return SYrequest;
        }
        #endregion

        #region Release the request of speech synthesizer
        /// <summary>
        /// 销毁语音合成对象.
        /// </summary>
        /// <param name="request">
        /// CreateSynthesizerRequest所建立的request对象.
        /// </param>
        /// <returns></returns>
        public void ReleaseSynthesizerRequest(SpeechSynthesizerRequest request)
        {
            NativeMethods.NlsReleaseSynthesizerRequest(request.native_request);
            request.native_request = IntPtr.Zero;
            return;
        }
        #endregion

        #region Creata a request of file transfer
        /// <summary>
        /// 创建录音文件识别对象.
        /// </summary>
        /// <returns></returns>
        public FileTransferRequest CreateFileTransferRequest()
        {
            IntPtr request = NativeMethods.NlsCreateFileTransferRequest();
            FileTransferRequest FTrequest = new FileTransferRequest();
            FTrequest.native_request = request;
            return FTrequest;
        }
        #endregion

        #region Release the request of file transfer
        /// <summary>
        /// 销毁录音文件识别对象.
        /// </summary>
        /// <param name="request">
        /// FileTransferRequest.
        /// </param>
        /// <returns></returns>
        public void ReleaseFileTransferRequest(FileTransferRequest request)
        {
            NativeMethods.NlsReleaseFileTransferRequest(request.native_request);
            request.native_request = IntPtr.Zero;
            return;
        }
        #endregion

        #region Create a NlsToken
        /// <summary>
        /// 创建Token获取对象.
        /// </summary>
        /// <returns>成功返回NlsToken对象, 否则返回NULL.</returns>
        public NlsToken CreateNlsToken()
        {
            IntPtr token = NativeMethods.NlsCreateNlsToken();
            NlsToken nlsToken = new NlsToken();
            nlsToken.native_token = token;
            return nlsToken;
        }
        #endregion

        #region Release the NlsToken
        /// <summary>
        /// 销毁Token获取对象.
        /// </summary>
        /// <param name="token">
        /// CreateNlsToken所建立的NlsToken对象.
        /// </param>
        /// <returns></returns>
        public void ReleaseNlsToken(NlsToken token)
        {
            NativeMethods.NlsReleaseNlsToken(token.native_token);
            token.native_token = IntPtr.Zero;
            return;
        }
        #endregion
    }
}
